File: 134__Gas_Station.py
class Solution(object):
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        dp = [gas[0] - cost[0]]
        for i in range(1, len(gas)):
            dp.append(dp[-1] + gas[i] - cost[i])
        if dp[-1] < 0:
            return -1
        minval, result = dp[0], 0
        for i in range(len(dp)):
            if minval > dp[i]:
                minval = dp[i]
                result = i
        return (result + 1) % len(gas)

File: Java/134__Gas_Station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int len = gas.length;
        int start = 0;
        int total_tank = 0, curr_tank = 0;
        for (int i = 0; i < len; i++) {
            total_tank += gas[i] - cost[i];
            curr_tank += gas[i] - cost[i];
            if (curr_tank < 0) {
                start = i + 1;
                curr_tank = 0;
            }
        }
        return total_tank >= 0 ? start : -1;
    }
}
